sootup.java.bytecode.frontend/src/main/java/sootup/java/bytecode/frontend/inputlocation/JrtFileSystemAnalysisInputLocation.java

```diff
@@ -149,21 +150,25 @@ protected Stream<JavaSootClassSource> getClassSourcesInternal(
             + classProvider.getHandledFileType().getExtensionWithDot();
 
     final Path archiveRoot = theFileSystem.getPath("modules", moduleSignature.getModuleName());
-    try {
-
-      return Files.walk(archiveRoot)
-          .filter(
-              filePath ->
-                  !Files.isDirectory(filePath)
-                      && filePath
-                          .toString()
-                          .endsWith(classProvider.getHandledFileType().getExtensionWithDot())
-                      && !filePath.toString().endsWith(moduleInfoFilename))
-          .flatMap(
-              p ->
-                  StreamUtils.optionalToStream(
-                      classProvider.createClassSource(this, p, fromPath(p, identifierFactory))))
-          .map(src -> (JavaSootClassSource) src);
+    try (Stream<Path> paths = Files.walk(archiveRoot)) {
+      // collect into a list and then return a stream, so we do not leak the Stream returned by
+      // Files.walk
+      List<JavaSootClassSource> javaSootClassSources =
+          paths
+              .filter(
+                  filePath ->
+                      !Files.isDirectory(filePath)
+                          && filePath
+                              .toString()
+                              .endsWith(classProvider.getHandledFileType().getExtensionWithDot())
+                          && !filePath.toString().endsWith(moduleInfoFilename))
+              .<SootClassSource>flatMap(
+                  p ->
+                      StreamUtils.optionalToStream(
+                          classProvider.createClassSource(this, p, fromPath(p, identifierFactory))))
+              .map(src -> (JavaSootClassSource) src)
+              .collect(Collectors.toList());
+      return javaSootClassSources.stream();
```

Comment:
please don' t .collect() and .stream() again

Reply:
Hi @swissiety since streams are lazy, if we don't collect, we cannot safely close the `Stream` returned by `Files.walk` within this method (since the filter and map operations won't have been executed yet).  The `collect` forces those operations to execute, so then the stream can be closed.  If you prefer not to collect and then stream again, an alternative would be to require all callers of this method to close the returned stream.  Would that be preferable?

Reply:
thx @msridhar for pointing to that situation! As this is a protected method and its calling contexts are collecting the Stream again i would prefer to close the Stream in the two calling method(s) due to have less allocations

Reply:
@Bryce7832 could you update the PR to get rid of the collect to a list, and instead use try-with-resources at both callers of this method?